index.js

```diff
@@ -1,44 +1,89 @@
+"use strict";
+
 var P = require('bluebird');
 var TokenFacilitator = require('token-facilitator');
-var crypto = require('crypto');
 var debug = require('debuglog')('hapi-stateless-notifications');
+var url = require('url');
 
 exports.register = function(server, options, next) {
   options = options || {};
 
+  debug("Registering plugin");
+
   server.ext('onPreHandler', function(request, reply) {
 
     if (!request.redis) {
       throw new Error("request.redis is not defined; set up a plugin that provides that");
     }
 
     request.saveNotifications = function(promises) {
-      return P.all(promises.map(function(promise) {
-        return P.resolve(promise).then(function(successNotice) {
-          debug("Success '%s' for request '%s'", successNotice, request.id);
-          return P.resolve({
-            notice: successNotice,
-            type: 'success'
-          });
-        }).catch(function(error) {
-          debug("Error '%s' for request '%s'", error.message, request.id);
-          return P.resolve({
-            notice: error.message,
-            type: 'error'
-          });
-        });
-      })).then(putNoticesInRedis(request.redis, options)).then(function (token) {
-        debug("Saved to redis for '%s' with token '%s'", request.id, token);
-        return token;
-      }, function (err) {
-        debug("Error saving to redis for '%s'", request.id);
-        throw err;
+      // Legacy API
+      return reply.saveNotifications(promises).then(function (data) {
+        return data.token
       });
     };
 
     return reply.continue();
   });
 
+  server.decorate('reply', 'saveNotifications', function (promises) {
+    var self = this;
+
+    return P.all(promises.map(function(promise) {
+      return P.resolve(promise).then(function(successNotice) {
+        debug("Success '%s' for request '%s'", successNotice, self.request.id);
+        return P.resolve({
+          notice: successNotice,
+          type: 'success'
+        });
+      }).catch(function(error) {
+        if ((!error.statusCode && error.constructor != Error) || error.statusCode >= 500) {
+          throw error;
+        }
+        debug("Error '%s' for request '%s'", error.message, self.request.id);
+        return P.resolve({
+          notice: error.message,
+          type: 'error'
+        });
+      });
+    })).then(function (notices) {
+      var anyFailed = Boolean(notices.find(function (n) {
```

Comment:
Yep. Though I haven't bumped the main body of code to ES6, so I left it using function(n) now.